File: since2020/UVa/UVa 11456.cpp
/**
 *  @judge UVa
 *  @id 11456
 *  @name Trainsorting
 * 
 *  @tag Longest Increasing Subsequence
 */
#include <bits/stdc++.h>
using namespace std;

typedef long long int LL;

int n;
vector<int> arr, lis, lds;

int solve() {
	lis = vector<int>(n, 1);
	lds = vector<int>(n, 1);
	for (int i = n - 1; i >= 0; i--)
		for (int j = i + 1; j < n; j++)
			if (arr[i] < arr[j])
				lds[i] = max(lds[i], lds[j] + 1);
			else if (arr[i] > arr[j])
				lis[i] = max(lis[i], lis[j] + 1);
	int ans = 0;
	for (int i = 0; i < n; i++)
		ans = max(ans, lis[i] + lds[i] - 1);
	lis.clear();
	lds.clear();
	return ans;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);
	int t;
	for (cin >> t; t; t--) {
		cin >> n;
		arr = vector<int>(n);
		for (auto && x: arr)
			cin >> x;
		cout << solve() << endl;
		arr.clear();
	}
}

File: since2020/UVa/UVa 11456.py
'''
	@judge UVa
	@id 11456
	@name Trainsorting

	@tag Longest Increasing Subsequence, Monotonicity
'''
from sys import stdin
from bisect import bisect_left

def solve(arr, n):
	if n == 0:
		return 0
	ls = arr[::-1] + arr
	lis = []
	for x in ls:
		if not lis or lis[-1] < x:
			lis.append(x)
		else:
			lis[ bisect_left(lis, x) ] = x
	return len(lis)

t = int(input())
while t > 0:
	n = int(input())
	arr = [ int(input()) for _ in range(n) ]
	print(solve(arr, n))
	t -= 1